<a href="https://colab.research.google.com/github/Trantracy/Python-practice/blob/master/BasicPython_Exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://i.imgur.com/0AUxkXt.png)

# Exercise 01 - Syntax, Variables and Numbers

Welcome to your first set of Python coding problems! 

**Notebooks** are composed of blocks (called "cells") of text and code. Each of these is editable, though you'll mainly be editing the code cells to answer some questions.

To get started, try running the code cell below (by pressing the `►| Run` button, or clicking on the cell and pressing `ctrl+Enter`/`shift+Enter` on your keyboard).

In [0]:
print("You've successfully run some Python code")
print("Congratulations!")

## 0. Creating a Variable

**What is your favorite color?**

To complete this question, create a variable called `color` in the cell below with an appropriate `string` value.

In [0]:
# Create a variable called color with an appropriate value on the line below
# (Remember, strings in Python must be enclosed in 'single' or "double" quotes)

color = "blue"

## 1. Simple Arithmetic Operation

Complete the code below. In case it's helpful, here is the table of available arithmetic operations:



| Operator     | Name           | Description                                            |
|--------------|----------------|--------------------------------------------------------|
| ``a + b``    | Addition       | Sum of ``a`` and ``b``                                 |
| ``a - b``    | Subtraction    | Difference of ``a`` and ``b``                          |
| ``a * b``    | Multiplication | Product of ``a`` and ``b``                             |
| ``a / b``    | True division  | Quotient of ``a`` and ``b``                            |
| ``a // b``   | Floor division | Quotient of ``a`` and ``b``, removing fractional parts |
| ``a % b``    | Modulus        | Integer remainder after division of ``a`` by ``b``     |
| ``a ** b``   | Exponentiation | ``a`` raised to the power of ``b``                     |
| ``-a``       | Negation       | The negative of ``a``                                  |

<span style="display:none"></span>


In [0]:
pi = 3.14159 # approximate
diameter = 3

# Create a variable called 'radius' equal to half the diameter
radius = diameter / 2
# Create a variable called 'area', using the formula for the area of a circle: pi times the radius squared
area = pi * (radius ** 2)

print(radius)
print(area)

1.5
7.0685775


**Results**:
- Area = 7.0685775

## 2. Variable Reassignment

Add code to the following cell to swap variables `a` and `b` (so that `a` refers to the object previously referred to by `b` and vice versa).

In [0]:
# If you're curious, these are examples of lists. We'll talk about 
# them in depth a few lessons from now. For now, just know that they're
# yet another type of Python object, like int or float.
a = [1, 2, 3]
b = [3, 2, 1]

######################################################################

# Your code goes here. Swap the values to which a and b refer.
# Hint: Try using a third variable
a, b = b, a

print(a, b)

[3, 2, 1] [1, 2, 3]


## 3. Order of Operations


a) Add parentheses to the following expression so that it evaluates to 1.

*Hint*: Following its default "PEDMAS"-like rules for order of operations, Python will first divide 3 by 2, then subtract the result from 5. You need to add parentheses to force it to perform the subtraction first.

In [0]:
(5 - 3) // 2

1

Questions, like this one, marked a spicy pepper are a bit harder. Don't feel bad if you can't get these.

b) <span title="A bit spicy" style="color: darkgreen ">🌶️</span> Add parentheses to the following expression so that it evaluates to 0

In [0]:
8 - 3 * 2 - (1 + 1)

0

## 4. Your Turn
Alice, Bob and Carol have agreed to pool their Halloween candies and split it evenly among themselves.
For the sake of their friendship, any candies left over will be smashed. For example, if they collectively
bring home 91 candies, they'll take 30 each and smash 1.

Write an arithmetic expression below to calculate how many candies they must smash for a given haul.

> *Hint*: You'll probably want to use the modulo operator, `%`, to obtain the remainder of division.

In [0]:
# Variables representing the number of candies collected by Alice, Bob, and Carol
alice_candies = 121
bob_candies = 77
carol_candies = 109

# Your code goes here! Replace the right-hand side of this assignment with an expression
# involving alice_candies, bob_candies, and carol_candies
to_smash = (alice_candies + bob_candies + carol_candies) % 3
to_smash

1

# Exercise 02 - Functions and Getting Help

## 1. Complete Your Very First Function 

Complete the body of the following function according to its docstring.

*HINT*: Python has a builtin function `round`

In [0]:
def round_to_two_places(num):
    """Return the given number rounded to two decimal places. 
    
    >>> round_to_two_places(3.14159)
    3.14
    """
    # Replace this body with your own code.
    # ("pass" is a keyword that does literally nothing. We used it as a placeholder,
    # so that it will not raise any errors,
    # because after we begin a code block, Python requires at least one line of code)
    return round(num, 2)

In [0]:
round_to_two_places(3.456)

3.46

## 2. Explore the Built-in Function
The help for `round` says that `ndigits` (the second argument) may be negative.
What do you think will happen when it is? Try some examples in the following cell?

Can you think of a case where this would be useful?

In [0]:
# Put your test code here
round(170.123, -2)

200.0

## 3. More Function

Giving the problem of candy-sharing friends Alice, Bob and Carol tried to split candies evenly. For the sake of their friendship, any candies left over would be smashed. For example, if they collectively bring home 91 candies, they'll take 30 each and smash 1.

Below is a simple function that will calculate the number of candies to smash for *any* number of total candies.

**Your task**: 
- Modify it so that it optionally takes a second argument representing the number of friends the candies are being split between. If no second argument is provided, it should assume 3 friends, as before.
- Update the docstring to reflect this new behaviour.

In [0]:
def to_smash(total_candies, nfriends=3):
    """Return the number of leftover candies that must be smashed after distributing
    the given number of candies evenly between 3 friends.
    
    >>> to_smash(91)
    1
    """
    return total_candies % nfriends

## 4. Taste some Errors

It may not be fun, but reading and understanding **error messages** will help you improve solving problem skills.

Each code cell below contains some commented-out buggy code. For each cell...

1. Read the code and predict what you think will happen when it's run.
2. Then uncomment the code and run it to see what happens. *(**Tip**: In the kernel editor, you can highlight several lines and press `ctrl`+`/` to toggle commenting.)*
3. Fix the code (so that it accomplishes its intended purpose without throwing an exception)

<!-- TODO: should this be autochecked? Delta is probably pretty small. -->

In [0]:
round_to_two_places(9.9999)

10.0

In [0]:
x = -10
y = 5
# Which of the two variables above has the smallest absolute value?
smallest_abs = min(abs(x), abs(y))

In [0]:
def f(x):
    y = abs(x)
    return y

print(f(5))

5


## 5. More and more Functions

For this question, we'll be using two functions imported from Python's `time` module.


### Time Function
The [time](https://docs.python.org/3/library/time.html#time.time) function returns the number of seconds that have passed since the Epoch (aka [Unix time](https://en.wikipedia.org/wiki/Unix_time)). 

<!-- We've provided a function called `seconds_since_epoch` which returns the number of seconds that have passed since the Epoch (aka [Unix time](https://en.wikipedia.org/wiki/Unix_time)). -->

Try it out below. Each time you run it, you should get a slightly larger number.

In [0]:
# Importing the function 'time' from the module of the same name. 
# (We'll discuss imports in more depth later)
from time import time
t = time()
print(t, "seconds since the Epoch")

1573637082.7703424 seconds since the Epoch


### Sleep Function
We'll also be using a function called [sleep](https://docs.python.org/3/library/time.html#time.sleep), which makes us wait some number of seconds while it does nothing particular. (Sounds useful, right?)

You can see it in action by running the cell below:

In [0]:
from time import sleep
duration = 5
print("Getting sleepy. See you in", duration, "seconds")
sleep(duration)
print("I'm back. What did I miss?")

Getting sleepy. See you in 5 seconds
I'm back. What did I miss?


### Your Own Function
With the help of these functions, complete the function **`time_call`** below according to its docstring.

<!-- (The sleep function will be useful for testing here since we have a pretty good idea of what something like `time_call(sleep, 1)` should return.) -->

In [0]:
def time_call(fn, arg):
    """Return the amount of time the given function takes (in seconds) when called with the given argument.
    """
    tic = time()
    fn(arg)
    toc = time()
    return toc-tic

How would you verify that `time_call` is working correctly? Think about it...

## 6. 🌶️

*Note: this question depends on a working solution to the previous question.*

Complete the function below according to its docstring.

In [0]:
def slowest_call(fn, arg1, arg2, arg3):
    """Return the amount of time taken by the slowest of the following function
    calls: fn(arg1), fn(arg2), fn(arg3)
    
    """
    return max(time_call(fn,arg1), time_call(fn,arg2), time_call(fn,arg3))

# Exercise 03 - Booleans and Conditionals

## 1. Simple Function with Conditionals

Many programming languages have [sign](https://en.wikipedia.org/wiki/Sign_function) available as a built-in function. Python does not, but we can define our own!

In the cell below, define a function called `sign` which takes a numerical argument and returns -1 if it's negative, 1 if it's positive, and 0 if it's 0.

In [0]:
# Your code goes here. Define a function called 'sign'
def sign(n):
    if n==0:
        return 0
    return -1 if n<0 else 1

In [0]:
sign(-10)

-1

## 2. Singular vs Plural Nouns

We've decided to add "print" to our `to_smash` function from Exercise 02

In [0]:
def to_smash(total_candies):
    """Return the number of leftover candies that must be smashed after distributing
    the given number of candies evenly between 3 friends.
    
    >>> to_smash(91)
    1
    """
    print("Splitting", total_candies, "candies")
    return total_candies % 3

to_smash(91)

Splitting 91 candies


1

What happens if we call it with `total_candies = 1`?

In [0]:
to_smash(1)

Splitting 1 candies


1

**Wrong grammar there!**

Modify the definition in the cell below to correct the grammar of our print statement.

**Your Task:**
> If there's only one candy, we should use the singular "candy" instead of the plural "candies"

In [0]:
def to_smash(total_candies):
    """Return the number of leftover candies that must be smashed after distributing
    the given number of candies evenly between 3 friends.
    
    >>> to_smash(91)
    1
    """
    print("Splitting", total_candies, 
          f"{'candy' if total_candies==1 else 'candies'}"
    )
    return total_candies % 3

to_smash(91)
to_smash(1)

Splitting 91 candies
Splitting 1 candy


1

## 3. Checking weather again

In the main lesson we talked about deciding whether we're prepared for the weather. I said that I'm safe from today's weather if...
- I have an umbrella...
- or if the rain isn't too heavy and I have a hood...
- otherwise, I'm still fine unless it's raining *and* it's a workday

The function below uses our first attempt at turning this logic into a Python expression. I claimed that there was a bug in that code. Can you find it?

To prove that `prepared_for_weather` is buggy, come up with a set of inputs where it returns the wrong answer.

In [0]:
def prepared_for_weather(have_umbrella, rain_level, have_hood, is_workday):
    # Don't change this code. Our goal is just to find the bug, not fix it!
    return have_umbrella or rain_level < 5 and have_hood or not (rain_level > 0 and is_workday)

# Change the values of these inputs so they represent a case where prepared_for_weather
# returns the wrong answer.
have_umbrella = False
rain_level = 0.0
have_hood = False
is_workday = False

# Check what the function returns given the current values of the variables above
actual = prepared_for_weather(have_umbrella, rain_level, have_hood, is_workday)
print(actual)

True


## 4. Start being lazy...

The function `is_negative` below is implemented correctly 
- It returns True if the given number is negative and False otherwise.

However, it's more verbose than it needs to be. We can actually reduce the number of lines of code in this function by *75%* while keeping the same behaviour. 

**Your task:**
> See if you can come up with an equivalent body that uses just **one line** of code, and put it in the function `concise_is_negative`. (HINT: you don't even need Python's ternary syntax)

In [0]:
def is_negative(number):
    if number < 0:
        return True
    else:
        return False

def concise_is_negative(number):
    return number < 0

## 5. Adding Toppings

The boolean variables `ketchup`, `mustard` and `onion` represent whether a customer wants a particular topping on their hot dog. We want to implement a number of boolean functions that correspond to some yes-or-no questions about the customer's order. For example:

In [0]:
def onionless(ketchup, mustard, onion):
    """Return whether the customer doesn't want onions.
    """
    return not onion

**Your task:**
> For each of the remaining functions, fill in the body to match the English description in the docstring. 

In [0]:
def wants_all_toppings(ketchup, mustard, onion):
    """Return whether the customer wants "the works" (all 3 toppings)
    """
    return ketchup and mustard and onion


In [0]:
def wants_plain_hotdog(ketchup, mustard, onion):
    """Return whether the customer wants a plain hot dog with no toppings.
    """
    return not (ketchup or mustard or onion)

In [0]:
def exactly_one_sauce(ketchup, mustard, onion):
    """Return whether the customer wants either ketchup or mustard, but not both.
    (You may be familiar with this operation under the name "exclusive or")
    """
    return (ketchup + mustard) == 1


## 6. <span title="A bit spicy" style="color: darkgreen ">🌶️</span>

We’ve seen that calling `bool()` on an integer returns `False` if it’s equal to 0 and `True` otherwise. What happens if we call `int()` on a bool? Try it out in the notebook cell below.

Can you take advantage of this to write a succinct function that corresponds to the English sentence "Does the customer want exactly one topping?"?

In [0]:
def exactly_one_topping(ketchup, mustard, onion):
    """Return whether the customer wants exactly one of the three available toppings
    on their hot dog.
    """
    return (ketchup + mustard + onion) == 1


# Exercises 04 - Lists

## 1. Second Element

Complete the function below according to its docstring.

*HINT*: Python starts counting at 0.

In [0]:
def select_second(L):
    """Return the second element of the given list. 
    If the list has no second element, return None.
    """
    return None if len(L)<2 else L[1]

## 2. Captain of the Worst Team

You are analyzing sports teams.  Members of each team are stored in a list. The Coach is the first name in the list, the Captain is the second name in the list, and other players are listed after that. 
These lists are stored in another list, which starts with the best team and proceeds through the list to the worst team last.  Complete the function below to select the **captain** of the worst team.

In [0]:
def losing_team_captain(teams):
    """Given a list of teams, where each team is a list of names, return the 2nd player (captain)
    from the last listed team
    """
    return teams[-1][1]

## 3. Purple Shell item

The next iteration of Mario Kart will feature an extra-infuriating new item, the ***Purple Shell***. When used, it warps the last place racer into first place and the first place racer into last place. Complete the function below to implement the Purple Shell's effect.

In [0]:
def purple_shell(racers):
    """Given a list of racers, set the first place racer (at the front of the list) to last
    place and vice versa.
    
    >>> r = ["Mario", "Bowser", "Luigi"]
    >>> purple_shell(r)
    >>> r
    ["Luigi", "Bowser", "Mario"]
    """
    racers[0], racers[-1] = racers[-1], racers[0]
    return racers

purple_shell(["Luigi", "Bowser", "Mario"])

['Mario', 'Bowser', 'Luigi']

In [0]:
a = [1, 2, 3]
def add(L):
    for i in range(len(L)):
        L[i] += 1
    print(L)
add(a)
a

b = 1
def add2(x):
    x += 1
    print(x)
add2(b)
b

[2, 3, 4]
2


1

## 4. Guess the Length!

What are the lengths of the following lists? Fill in the variable `lengths` with your predictions. (Try to make a prediction for each list *without* just calling `len()` on it.)

In [0]:
a = [1, 2, 3]
b = [1, [2, 3]]
c = []
d = [1, 2, 3][1:]

# Put your predictions in the list below. Lengths should contain 4 numbers, the
# first being the length of a, the second being the length of b and so on.
lengths = [3, 2, 0, 2]

## 5. <span title="A bit spicy" style="color: darkgreen ">🌶️</span>

We're using lists to record people who attended our party and what order they arrived in. For example, the following list represents a party with 7 guests, in which Adela showed up first and Ford was the last to arrive:

    party_attendees = ['Adela', 'Fleda', 'Owen', 'May', 'Mona', 'Gilbert', 'Ford']

A guest is considered 'fashionably late' if they arrived after at least half of the party's guests. However, they must not be the very last guest (that's taking it too far). In the above example, Mona and Gilbert are the only guests who were fashionably late.

Complete the function below which takes a list of party attendees as well as a person, and tells us whether that person is fashionably late.

In [0]:
def fashionably_late(arrivals, name):
    """Given an ordered list of arrivals to the party and a name, return whether the guest with that
    name was fashionably late.
    """
    return (arrivals.index(name) > len(arrivals)//2) and (arrivals.index(name) != len(arrivals)-1)


# Exercises 05 - Loops and List Comprehensions

## 1. Lucky number

Complete the body of the function below according to its docstring.

In [0]:
def has_lucky_number(nums):
    """Return whether the given list of numbers is lucky. A lucky list contains
    at least one number divisible by 7.
    """
    return sum([True if x % 7 == 0 else False for x in nums]) > 0

has_lucky_number([71, 141, 2, 3, 4])

False

In [0]:
not all([False, True, True])

True

## 2. Upgrade the List



### a. Guess the Error
Look at the Python expression below. What do you think we'll get when we run it? When you've made your prediction, uncomment the code and run the cell to see if you were right.

In [0]:
[1, 2, 3, 4] > 2

### b. Upgrade the List
R and Python have some libraries (like Numpy and Pandas) compare each element of the list to 2 (i.e. do an 'element-wise' comparison) and give us a list of booleans like `[False, False, True, True]`. 

Implement a function that reproduces this behaviour, returning a list of booleans corresponding to whether the corresponding element is greater than n.

In [0]:
def elementwise_greater_than(L, thresh):
    """Return a list with the same length as L, where the value at index i is 
    True if L[i] is greater than thresh, and False otherwise.
    
    >>> elementwise_greater_than([1, 2, 3, 4], 2)
    [False, False, True, True]
    """
    return [x > thresh for x in L]

elementwise_greater_than([1, 2, 3, 4], 2)

[False, False, True, True]

## 3. Boring Menu

Complete the body of the function below according to its docstring.

*Hint*: This is a case where it may be preferable to iterate over the indices of the list (using a call to `range()`) rather than iterating over the elements of the list itself. 

In [0]:
def menu_is_boring(meals):
    """Given a list of meals served over some period of time, return True if the
    same meal has ever been served two days in a row, and False otherwise.
    """
    for i in range(len(meals)-1):
        if meals[i] == meals[i+1]:
            return True
    return False

# Exercises 06 - Strings and Dictionaries

## 0. Length of Strings

Let's start with a string lightning round to warm up. What are the lengths of the strings below?

For each of the five strings below, predict what `len()` would return when passed that string. Use the variable `length` to record your answer.

In [0]:
a = ""
length = 0

In [0]:
b = "it's ok"
length = 7

In [0]:
c = 'it\'s ok'
length = 7

In [0]:
d = """
hey
"""
length = 3

In [0]:
e = '\n'
length = 1

## 1. Check the Zip Code

There is a saying that *"Data scientists spend 80% of their time cleaning data, and 20% of their time complaining about cleaning data."* Let's see if you can write a function to help clean US zip code data. Given a string, it should return whether or not that string represents a valid zip code. For our purposes, a valid zip code is any string consisting of exactly 5 digits.

HINT: `str` has a method that will be useful here. Use `help(str)` to review a list of string methods.

In [0]:
def is_valid_zip(zip_code):
    """Returns whether the input string is a valid (5 digit) zip code
    """
    if len(zip_code) != 5:
        return False
    for i in zip_code:
        if not i.isdigit():
            return False
    return True

is_valid_zip('1234r')

False

## 2. Searching a Word

A researcher has gathered thousands of news articles. But she wants to focus her attention on articles including a specific word. Complete the function below to help her filter her list of articles.

Your function should meet the following criteria

- Do not include documents where the keyword string shows up only as a part of a larger word. For example, if she were looking for the keyword “closed”, you would not include the string “enclosed.” 
- She does not want you to distinguish upper case from lower case letters. So the phrase “Closed the case.” would be included when the keyword is “closed”
- Do not let periods or commas affect what is matched. “It is closed.” would be included when the keyword is “closed”. But you can assume there are no other types of punctuation.

*HINT*: Some methods that may be useful here: `str.split()`, `str.strip()`, `str.lower()`

In [0]:
def word_search(doc_list, keyword):
    """
    Takes a list of documents (each document is a string) and a keyword. 
    Returns list of the index values into the original list for all documents 
    containing the keyword.

    Example:
    doc_list = ["The Learn Python Challenge Casino.", "They bought a car", "Casinoville"]
    >>> word_search(doc_list, 'casino')
    >>> [0]
    """
    result = []
    for i in range(len(doc_list)):
        words = [word.lower().strip(',.') for word in doc_list[i].split()]
        if keyword in words:
            result.append(i)
    return result

doc_list = ["The Learn Python Challenge Casino.", "They bought a car", "Casinoville"]
word_search(doc_list, 'casino')

[0]

In [0]:
s = 'asdfdf!@#,..,'
import re

re.sub('\W+', '', s)

'asdfdf'

## 3. Searching Multiple Words

Now the researcher wants to supply multiple keywords to search for. Complete the function below to help her.

(You're encouraged to use the `word_search` function you just wrote when implementing this function. Reusing code in this way makes your programs more robust and readable - and it saves typing!)

In [0]:
def multi_word_search(doc_list, keywords):
    """
    Takes list of documents (each document is a string) and a list of keywords.  
    Returns a dictionary where each key is a keyword, and the value is a list of indices
    (from doc_list) of the documents containing that keyword

    >>> doc_list = ["The Learn Python Challenge Casino.", "They bought a car and a casino", "Casinoville"]
    >>> keywords = ['casino', 'they']
    >>> multi_word_search(doc_list, keywords)
    {'casino': [0, 1], 'they': [1]}
    """
    d = {}
    for keyword in keywords:
        d[keyword] = word_search(doc_list, keyword)
    return d

doc_list = ["The Learn Python Challenge Casino.", "They bought a car and a casino", "Casinoville"]
keywords = ['casino', 'they']
multi_word_search(doc_list, keywords)  

{'casino': [0, 1], 'they': [1]}

# The End

You've finished the Python course. Congrats!
Happy Pythoning!